featurebyte/routes/user_defined_function/controller.py

```diff
@@ -0,0 +1,226 @@
+from featurebyte.routes.common.base import BaseDocumentController
+from featurebyte.schema.user_defined_function import (
+    UserDefinedFunctionCreate,
+    UserDefinedFunctionList,
+    UserDefinedFunctionServiceUpdate,
+    UserDefinedFunctionUpdate,
+)
+from featurebyte.service.feature import FeatureService
+from featurebyte.service.feature_store import FeatureStoreService
+from featurebyte.service.feature_store_warehouse import FeatureStoreWarehouseService
+from featurebyte.service.user_defined_function import UserDefinedFunctionService
+
+
+class UserDefinedFunctionController(
+    BaseDocumentController[
+        UserDefinedFunctionModel, UserDefinedFunctionService, UserDefinedFunctionList
+    ]
+):
+    """
+    UserDefinedFunctionController class
+    """
+
+    paginated_document_class = UserDefinedFunctionList
+
+    def __init__(
+        self,
+        user_defined_function_service: UserDefinedFunctionService,
+        feature_service: FeatureService,
+        feature_store_service: FeatureStoreService,
+        feature_store_warehouse_service: FeatureStoreWarehouseService,
+    ):
+        super().__init__(service=user_defined_function_service)
+        self.feature_service = feature_service
+        self.feature_store_service = feature_store_service
+        self.feature_store_warehouse_service = feature_store_warehouse_service
+
+    async def create_user_defined_function(
+        self,
+        data: UserDefinedFunctionCreate,
+        get_credential: Any,
+    ) -> UserDefinedFunctionModel:
+        """
+        Create UserDefinedFunction at persistent
+
+        Parameters
+        ----------
+        data: UserDefinedFunctionCreate
+            UserDefinedFunction creation payload
+        get_credential: Any
+            Function to get credential
+
+        Returns
+        -------
+        UserDefinedFunctionModel
+            Newly created user_defined_function object
+
+        Raises
+        ------
+        DocumentCreationError
+            If user defined function not exists in warehouse
+        """
+        # validate feature store id exists
+        feature_store = await self.feature_store_service.get_document(
+            document_id=data.feature_store_id
+        )
+
+        # create user defined function
+        user_defined_function = await self.service.create_document(data)
+        try:
+            # check if user defined function exists in warehouse
+            await self.feature_store_warehouse_service.check_user_defined_function_exists(
+                user_defined_function=user_defined_function,
+                feature_store=feature_store,
+                get_credential=get_credential,
+            )
+        except Exception as exc:
+            # if not exists, delete the user defined function
+            await self.service.delete_document(document_id=user_defined_function.id)
+            raise DocumentCreationError(f"User defined function not exists: {exc}") from exc
```

Comment:
If the function exists but `get_default_test_value()` raises ValueError because the parameter type is not supported, will it be raised through here? Would be a bit nicer if the error message can reflect that instead of always mentioning the function does not exist.

Reply:
The error message will be kept inside `{exc}`. Will update this part to drop `User defined function not exists: `.